### PR TITLE
Db query metrics

### PR DESCRIPTION
**Pull request by dlanderson**
Initial database query metrics. WIP.

### CODE REVIEW

File: lib/scout_apm/db_query_metric_stats.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      self.min_call_time = other.min_call_time if self.min_call_time.zero? or other.min_call_time < self.min_call_time
+      self.max_call_time = other.max_call_time if other.max_call_time > self.max_call_time
+      self.histogram.combine!(other.histogram)
+      self
+    end
+
+    # To avoid conflicts with different JSON libaries handle JSON ourselves.
+    # Time-based metrics are converted to milliseconds from seconds.
+    def to_json(*a)
+       %Q[{"call_count":#{call_count},"total_call_time":#{call_time*1000},"min_call_time":#{min_call_time*1000},"max_call_time":#{max_call_time*1000},"rows_returned":#{rows_returned}}, "histogram":#{histogram.as_json}]
```
Comment (cschneid) [inline lib/scout_apm/db_query_metric_stats.rb]: Break this into several lines - no need to prematurely compress this string like this, we gzip it.

File: lib/scout_apm/layer_converters/database_converter.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        scope = scope_layer
+
+        # TODO: Track requests that never reach a Controller (for example, when
+        # Middleware decides to return rather than passing onward)
+        return {} unless scope
+
+        create_database_metrics
+      end
+
+      def create_database_metrics
```
Comment (cschneid) [inline lib/scout_apm/layer_converters/database_converter.rb]: This doesn't create metrics, it returns a list of layers

File: lib/scout_apm/serializers/payload_serializer_to_json.rb [hunk 3]
```diff
@@ -11,6 +11,7 @@ def serialize(metadata, metrics, slow_transactions, jobs, slow_jobs, histograms)
                         :jobs => JobsSerializerToJson.new(jobs).as_json,
                         :slow_jobs => SlowJobsSerializerToJson.new(slow_jobs).as_json,
                         :histograms => HistogramsSerializerToJson.new(histograms).as_json,
+                        :db_query_metrics => DbQuerySerializerToJson.new(db_query_metrics).as_json
```
Comment (cschneid) [inline lib/scout_apm/serializers/payload_serializer_to_json.rb]: trailing comma plz.

File: lib/scout_apm/store.rb [hunk 4]
```diff
@@ -23,29 +23,36 @@ def current_period
       reporting_periods[current_timestamp]
     end
 
+    def find_period(timestamp = nil)
+      if timestamp
+        @reporting_periods[options[:timestamp]]
```
Comment (cschneid) [inline lib/scout_apm/store.rb]: `options` doesn't exist here.

File: lib/scout_apm/db_query_metric_set.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+    def lookup(other)
+      metrics[other.key] ||= other
+    end
+
+    def combine!(other)
+      lookup(other).combine!(other)
+    end
+
+    # Takes an array of ActiveRecord layers, creates new DbQueryMetricStats and combines them into this Set
```
Comment (cschneid) [inline lib/scout_apm/db_query_metric_set.rb]: Not sure I like this class knowing so much about how we store layers. Elsewhere we stop caring that the data was in layers after the converter step, but in the db case it goes all the way through the Store, into this class.

Can we create those DbQueryMetricStats a lot earlier, and have this class take those as its input?

**Comment by joelr** [comment]
😬

**Pull request merged**